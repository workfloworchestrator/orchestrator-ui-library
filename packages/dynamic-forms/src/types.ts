import React, { FormEventHandler } from 'react';
import { FieldValues, useForm } from 'react-hook-form';

import { z } from 'zod';

export type TDynamicFormFieldOptionsType = 'anyOf' | 'allOf';

export enum DfFieldFormats {
    DATE = 'date',
    HIDDEN = 'hidden',
    DATETIME = 'date-time',
    SKIP = 'skip',
    TIMESTAMP = 'timestamp',

    DEFAULT = 'text',
    DROPDOWN = 'dropdown',
    BOOLFIELD = 'boolfield',
    SELECT = 'select',
    MULTISELECT = 'multiselect',
    CHECKBOX = 'checkbox',
    RADIO = 'radio',
    SWITCH = 'switch',
    DATEPICKER = 'datepicker',
    NUMBER = 'number',
    LIST = 'list',
}

export enum DfFieldTypes {
    OPTGROUP = 'optGroup',
    SKIP = 'skip',
    LONG = 'long',
    HIDDEN = 'hidden',
    LABEL = 'label',
    SUMMARY = 'summary',
    ACCEPT = 'accept',
    NULL = 'null',

    DATE = 'date',
    DATETIME = 'date-time',
    OBJECT = 'object',
    STRING = 'string',
    ARRAY = 'array',
    BOOLEAN = 'boolean',
    NUMBER = 'number',
    TIMESTAMP = 'timestamp',
    LIST = 'list',
    DEFINED_LIST = 'defined-list',
    DEFINED_MULTI_LIST = 'defined-multi-list',
    TWO_OPTION_LIST = 'two-options-list',
}

export enum DFFieldReturnType {
    STRING = 'string',
    NUMBER = 'number',
    DATETIME = 'date-time',
}
export interface IValidationErrorDetails {
    detail: string;
    source: IDynamicFormApiValidationError[];
    mapped: {
        [fieldId: string]: {
            value: string;
            msg: string;
        };
    };
}

export interface IDynamicFormsContextProps {
    isLoading: boolean;
    isSending: boolean;
    isFullFilled: boolean;
    theReactHookForm: ReturnType<typeof useForm>;
    errorDetails?: IValidationErrorDetails;
    formData?: IDynamicForm;
    debugMode?: boolean;
    title?: string;
    sendLabel?: string;
    onCancel?: () => void;
    submitForm: FormEventHandler<HTMLFormElement>;
    resetForm: (e: MouseEvent) => void;
    successNotice?: React.ReactNode;
    headerComponent?: React.ReactNode;
    footerComponent?: React.ReactNode;
}

export type DynamicFormsMetaData = {
    [key: string | number]: DfFieldValue;
};

export type DfFormProvider = ({
    formKey,
    requestBody,
}: {
    formKey: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    requestBody: any;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Promise<Record<string, any>>;

export type DfLabelProvider = ({
    formKey,
    id,
}: {
    formKey: string;
    id?: string | null;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) => Promise<Record<string, any>>;

export type DfDataProvider = () => Promise<IDynamicFormsLabels>;

export type CustomValidationRuleFn = (
    fieldConfig: IDynamicFormField,
    theReactHookForm?: ReturnType<typeof useForm>,
) => Zod.ZodTypeAny | undefined;

export interface IDynamicFormsContextConfig {
    // use custom method to provide data for the form. This overwrites data fetched from labels endpoint
    dataProvider: DfDataProvider;

    // use custom method for providing labels and data
    labelProvider: DfLabelProvider;

    // use a custom method for providing the form definition
    formProvider: DfFormProvider;

    // Extend field definitions
    fieldDetailProvider?: IFieldDefinitionProvider;

    // be able to refresh the provided data
    dataProviderCacheKey?: number;

    // whenever a fieldvalue changes, do something
    onFieldChangeHandler?: onFieldChangeHandlerFn;

    // provide custom validation rules for fields
    customValidationRules?: CustomValidationRuleFn;

    // This is a temp solution for 2 different Backend implementations of pydantic forms
    // we will remove this once the backends both work the same
    // Portal backend has a hack that is a leftover from previous implementation
    // So for only that backend, we should not send the data as a array
    tmp_pydanticFormsOriginalImplementation?: boolean;
}

export interface IDynamicFormsContextInitialProps {
    formKey: string;
    formIdKey?: string;
    title?: string;
    sendLabel?: string;
    metaData?: DynamicFormsMetaData;
    successNotice?: React.ReactNode;
    onSuccess?: (fieldValues: FieldValues) => void;
    onCancel?: () => void;
    children: (props: IDynamicFormsContextProps) => React.ReactNode;
    headerComponent?: React.ReactNode;
    footerComponent?: React.ReactNode;

    config: IDynamicFormsContextConfig;
}

export interface IFieldDefinitionProvider {
    [fieldId: string]: Partial<IDynamicFormField>;
}

export type onFieldChangeHandlerFn = (
    field: {
        name?: string;
        type?: string;
        value: DfFieldValue;
    },
    theReactHookForm: ReturnType<typeof useForm>,
) => void;

export interface IDynamicFormApiResponseDefEnum {}

export interface TJsonSchemaRef {
    $ref: string;
}

export interface TDynamicFormFieldAnyOfDef {
    items?: TJsonSchemaRef;
    type: 'null' | 'array';
}

export type ListFieldType =
    | DfFieldTypes.STRING
    | DfFieldTypes.DATE
    | DfFieldTypes.NUMBER
    | DfFieldTypes.NULL;

export interface ITDynamicFormFieldAnyOfResolvedItems {
    enum: string[];
    title: string;
    type: ListFieldType;
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface TDynamicFormFieldAnyOfResolved {
    items?: ITDynamicFormFieldAnyOfResolvedItems;
    enum?: string[];
    format?: 'date' | 'date-time';
    type: ListFieldType;
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface IDynamicFormApiResponseProperty {
    type: DfFieldTypes;
    anyOf?: TDynamicFormFieldAnyOfDef[];
    default?: string | null;
    title: string;
}

export interface IDynamicFormApiResponsePropertyResolved {
    type: DfFieldTypes;

    anyOf?: TDynamicFormFieldAnyOfResolved[];
    oneOf?: TDynamicFormFieldAnyOfResolved[];
    allOf?: TDynamicFormFieldAnyOfResolved[];

    items?: ITDynamicFormFieldAnyOfResolvedItems;
    enum?: string[];

    default?: string | null;
    title: string;
    format: DfFieldFormats;

    uniforms?: {
        disabled: boolean;
    };

    // validation props
    maxLength?: number;
    minLength?: number;
    pattern?: string;
}

export interface IDynamicFormApiResponseBase {
    title: string;
    description: string;
    additionalProperties: boolean;
    type: 'object';
    required?: string[];
    $defs?: {
        [definitionId: string]: {
            enum: string[];
            title: string;
            type: DfFieldTypes;
        };
    };
}

export interface IDynamicFormApiResponse extends IDynamicFormApiResponseBase {
    properties: {
        [propId: string]: IDynamicFormApiResponseProperty;
    };
}

export interface IDynamicFormApiErrorResponse {
    detail?: string;
    status: number;
    form: IDynamicFormApiResponse;
    success?: boolean;
    validation_errors: IDynamicFormApiValidationError[];
}

export interface IDynamicFormApiValidationError {
    input: string;
    loc: string[];
    msg: string;
    type: string;
    url: string; //"https://errors.pydantic.dev/2.4/v/extra_forbidden"
}

export interface IDynamicFormApiRefResolved
    extends IDynamicFormApiResponseBase {
    properties: {
        [propId: string]: IDynamicFormApiResponsePropertyResolved;
    };
}

export interface IDynamicFormsLabels {
    [key: string]: string[] | number[] | string | number | null;
}

export interface IDynamicFormsLabelResponse {
    labels: IDynamicFormsLabels;
}

export interface IDynamicFormFieldValidation {
    maxLength?: number;
    minLength?: number;
    pattern?: string;
    isNullable?: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type DfFieldValue = any;

export interface IDynamicFormFieldAttributes {
    disabled?: boolean;
}

export interface IDynamicFormField {
    id: string;
    title: string;
    description: string;
    type: DfFieldTypes;
    format: DfFieldFormats;
    options: string[];
    disabledOptions?: string[];
    default?: DfFieldValue;
    required: boolean;
    isEnumField: boolean;
    schemaField: IDynamicFormApiResponsePropertyResolved;
    validation: IDynamicFormFieldValidation;
    attributes: IDynamicFormFieldAttributes;
    validator?: FormZodValidationFn;
    FormElement?: FormElementComponent;
    matchedFieldResult?: DfFieldConfig;
}

export interface IDynamicFormFieldSection {
    id: string;
    title: string;
    fields: string[];
}

export enum IDynamicFormState {
    NEW = 'new',
    IN_PROGRESS = 'in-progress',
    FINISHED = 'finished',
}

export interface IDynamicForm {
    title: string;
    description: string;
    state: IDynamicFormState;
    fields: IDynamicFormField[];
    sections: IDynamicFormFieldSection[];
}

export type FormZodValidationFn = (field: IDynamicFormField) => z.ZodTypeAny;
export type FormElementComponent = (props: IDFInputFieldProps) => JSX.Element;

export interface FormComponent {
    Element: FormElementComponent;
    validator?: FormZodValidationFn;
}

export interface DfFieldConfig {
    id: string;
    Component: FormComponent;
    matcher?: (field: IDynamicFormField) => boolean;
    preventColRender?: boolean;
}

export type DfFieldsConfig = DfFieldConfig[];

export interface IDFInputFieldProps {
    field: IDynamicFormField;
}

export interface IDFZodValidationPresets {
    [type: string]: FormZodValidationFn;
}
