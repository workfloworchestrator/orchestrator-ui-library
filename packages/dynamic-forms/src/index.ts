import DynamicForm from "./components/form/ContextWrap"

export default DynamicForm
