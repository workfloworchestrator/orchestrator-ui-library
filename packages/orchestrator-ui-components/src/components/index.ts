export * from './WfoBadges';
export * from './WfoPageTemplate/WfoBreadcrumbs';
export * from './WfoFilterTabs';
export * from './WfoKeyValueTable';
export * from './WfoPageTemplate';
export * from './WfoSearchBar';
export * from './WfoSettingsModal';
export * from './WfoSettingsPage';
export * from './WfoStartPage';
export * from './WfoSubscription';
export * from './WfoTable';
export * from './WfoTimeline';
export * from './WfoLoading';
export * from './WfoToastsList';
export * from './WfoForms';
export * from './WfoStartTaskButtonComboBox';
