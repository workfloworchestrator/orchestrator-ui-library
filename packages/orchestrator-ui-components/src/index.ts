export * from './lib/components';

export * from './lib/icons';

export * from './lib/theme';

export * from './lib/utils';

export * from './lib/contexts';

export * from './lib/types';

export * from './lib/hooks';
