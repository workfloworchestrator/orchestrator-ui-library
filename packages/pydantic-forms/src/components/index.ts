export * from './forms';
export * from './formFields';
