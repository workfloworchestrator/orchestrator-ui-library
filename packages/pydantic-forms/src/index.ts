export * from './components/index';
export * from './contexts/index';
export * from './messages/index';
export * from './types/index';
