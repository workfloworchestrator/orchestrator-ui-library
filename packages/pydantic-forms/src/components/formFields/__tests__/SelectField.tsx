/*
 * Copyright 2019-2023 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import waitForComponentToPaint from "__tests__/waitForComponentToPaint";
import createContext from "lib/uniforms-surfnet/__tests__/_createContext";
import mount from "lib/uniforms-surfnet/__tests__/_mount";
import { ListField, SelectField } from "lib/uniforms-surfnet/src";
import React from "react";
import ReactSelect from "react-select";

describe("<SelectField>", () => {
    test("<SelectField> - renders an input", () => {
        const element = <SelectField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
    });

    test("<SelectField> - renders a select with correct disabled state", () => {
        const element = <SelectField name="x" disabled />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("isDisabled")).toBe(true);
    });

    test("<SelectField> - renders a select with correct readOnly state", () => {
        const onChange = jest.fn();

        const element = <SelectField name="x" readOnly />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }, { onChange }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        wrapper.find(ReactSelect).invoke("onChange")({ value: "b" });
        expect(onChange).not.toHaveBeenCalled();
    });

    test("<SelectField> - renders a select with correct id (inherited)", () => {
        const element = <SelectField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("id")).toBeTruthy();
    });

    test("<SelectField> - renders a select with correct id (specified)", () => {
        const element = <SelectField name="x" id="y" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("id")).toBe("y");
    });

    test("<SelectField> - renders a select with correct name", () => {
        const element = <SelectField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("name")).toBe("x");
    });

    test("<SelectField> - renders a select with correct options", () => {
        const element = <SelectField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("options")).toStrictEqual([
            { label: "a", text: "a", value: "a" },
            { label: "b", text: "b", value: "b" },
        ]);
    });

    test("<SelectField> - renders a select with correct options (transform)", () => {
        const element = <SelectField name="x" transform={(x: string) => x.toUpperCase()} />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("options")).toStrictEqual([
            { label: "A", text: "A", value: "a" },
            { label: "B", text: "B", value: "b" },
        ]);
    });

    test("<SelectField> - renders a select with correct placeholder (fallback)", () => {
        const element = <SelectField name="x" placeholder="" />;
        const wrapper = mount(
            element,
            createContext({
                x: { type: String, allowedValues: ["a", "b"], optional: true },
            })
        );

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("options")).toStrictEqual([
            { label: "a", text: "a", value: "a" },
            { label: "b", text: "b", value: "b" },
        ]);
        expect(wrapper.find(ReactSelect).prop("placeholder")).toBe("Search and select a value...");
    });

    test("<SelectField> - renders a select with correct placeholder (implicit)", () => {
        const element = <SelectField name="x" placeholder="y" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("options")).toStrictEqual([
            { label: "a", text: "a", value: "a" },
            { label: "b", text: "b", value: "b" },
        ]);
        expect(wrapper.find(ReactSelect).prop("placeholder")).toBe("y");
    });

    test("<SelectField> - renders a select with correct value (default)", () => {
        const element = <SelectField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("value")).toBe(undefined);
    });

    test("<SelectField> - renders a select with correct value (model)", () => {
        const element = <SelectField name="x" id="snapshot-test" />;
        const wrapper = mount(
            element,
            createContext(
                { x: { type: String, allowedValues: ["a", "b"] } },
                { model: { x: "b" } },
                "selectfield-model"
            )
        );

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("value")).toStrictEqual({ label: "b", text: "b", value: "b" });
        expect(wrapper.render()).toMatchSnapshot();
    });

    test("<SelectField> - renders a select with correct value (specified)", () => {
        const element = <SelectField name="x" value="b" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        expect(wrapper.find(ReactSelect).prop("value")).toStrictEqual({ label: "b", text: "b", value: "b" });
    });

    test("<SelectField> - renders a select which correctly reacts on change", () => {
        const onChange = jest.fn();

        const element = <SelectField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }, { onChange }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        wrapper.find(ReactSelect).invoke("onChange")({ value: "b" });

        expect(onChange).toHaveBeenLastCalledWith("x", "b");
    });

    test("<SelectField> - renders a select which correctly reacts on change (empty)", () => {
        const onChange = jest.fn();

        const element = <SelectField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }, { onChange }));

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        wrapper.find(ReactSelect).invoke("onChange")(undefined);

        expect(onChange).toHaveBeenLastCalledWith("x", undefined);
    });

    test("<SelectField> - renders a select which correctly reacts on change (same value)", () => {
        const onChange = jest.fn();

        const element = <SelectField name="x" />;
        const wrapper = mount(
            element,
            createContext({ x: { type: String, allowedValues: ["a", "b"] } }, { model: { x: "b" }, onChange })
        );

        expect(wrapper.find(ReactSelect)).toHaveLength(1);
        wrapper.find(ReactSelect).invoke("onChange")({ value: "b" });

        expect(onChange).toHaveBeenLastCalledWith("x", "b");
    });

    test("<SelectField> - renders a label", () => {
        const element = <SelectField name="x" label="y" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find("label")).toHaveLength(1);
        expect(wrapper.find("label").prop("children")).toContain("y");
        expect(wrapper.find("label").prop("htmlFor")).toBe(wrapper.find(ReactSelect).prop("id"));
    });

    test("<SelectField> - renders a wrapper with unknown props", () => {
        const element = <SelectField name="x" data-x="x" data-y="y" data-z="z" />;
        const wrapper = mount(element, createContext({ x: { type: String, allowedValues: ["a", "b"] } }));

        expect(wrapper.find("section").at(0).prop("data-x")).toBe("x");
        expect(wrapper.find("section").at(0).prop("data-y")).toBe("y");
        expect(wrapper.find("section").at(0).prop("data-z")).toBe("z");
    });

    test("<SelectField> - renders a select with correct value (as uniqueItem list child)", async () => {
        const element = <ListField name="x" id="snapshot-test" />;
        const wrapper = mount(
            element,
            createContext(
                {
                    x: { type: Array, uniforms: { uniqueItems: true } },
                    "x.$": { type: String, allowedValues: ["a", "b"] },
                },
                { model: { x: ["a", undefined] } },
                "selectfield-uniqueitem"
            )
        );

        await waitForComponentToPaint(wrapper);

        expect(wrapper.find(ReactSelect)).toHaveLength(2);
        expect(wrapper.find(ReactSelect).at(0).prop("value")).toStrictEqual({ label: "a", text: "a", value: "a" });
        expect(wrapper.find(ReactSelect).at(1).prop("value")).toStrictEqual(undefined);
        expect(wrapper.find(ReactSelect).at(1).prop("options")).toStrictEqual([{ label: "b", text: "b", value: "b" }]);
        expect(wrapper.render()).toMatchSnapshot();
    });

    test("<SelectField> - works with special characters", () => {
        mount(<SelectField name="x" />, createContext({ x: { type: String, allowedValues: ["ă", "ș"] } }));
    });
});
