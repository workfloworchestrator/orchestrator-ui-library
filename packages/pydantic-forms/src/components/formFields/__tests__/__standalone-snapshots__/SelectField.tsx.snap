// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectField> <SelectField> - renders a select with correct value (as uniqueItem list child) 1`] = `
LoadedCheerio {
  "0": <div
    class="euiFlexItem css-ys1ii3-EuiFlexItem"
  >
    <section
      class="list-field"
      id="snapshot-test"
      required=""
    >
      <div
        class="euiFormRow euiFormRow--fullWidth"
        id="formrow-x-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            class="euiFormLabel euiFormRow__label"
            for="formrow-x"
            id="formrow-x-label"
          />
           
          <div
            class="euiText euiText--medium"
          />
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        />
      </div>
      <ul>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="selectfield-uniqueitem-0003-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="selectfield-uniqueitem-0003"
                  id="selectfield-uniqueitem-0003-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class=" css-b62m3t-container"
                  id="selectfield-uniqueitem-0003"
                >
                  <span
                    class="css-1f43avz-a11yText-A11yText"
                    id="react-select-20-live-region"
                  />
                  <span
                    aria-atomic="false"
                    aria-live="polite"
                    aria-relevant="additions text"
                    class="css-1f43avz-a11yText-A11yText"
                  />
                  <div
                    class=" css-j34sbj-control"
                  >
                    <div
                      class=" css-319lph-ValueContainer"
                    >
                      <div
                        class=" css-30t56x-singleValue"
                      >
                        a
                      </div>
                      <div
                        class=" css-16cuj0n-Input"
                        data-value=""
                      >
                        <input
                          aria-autocomplete="list"
                          aria-expanded="false"
                          aria-haspopup="true"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          class=""
                          id="selectfield-uniqueitem-0003.search"
                          role="combobox"
                          spellcheck="false"
                          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                      <span
                        class=" css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <input
                    name="x.0"
                    type="hidden"
                    value="a"
                  />
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="selectfield-uniqueitem-0005.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <li>
          <section>
            <div
              class="euiFormRow euiFormRow--fullWidth"
              id="selectfield-uniqueitem-0008-row"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  class="euiFormLabel euiFormRow__label"
                  for="selectfield-uniqueitem-0008"
                  id="selectfield-uniqueitem-0008-label"
                />
                 
                <div
                  class="euiText euiText--medium"
                />
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  class=" css-b62m3t-container"
                  id="selectfield-uniqueitem-0008"
                >
                  <span
                    class="css-1f43avz-a11yText-A11yText"
                    id="react-select-21-live-region"
                  />
                  <span
                    aria-atomic="false"
                    aria-live="polite"
                    aria-relevant="additions text"
                    class="css-1f43avz-a11yText-A11yText"
                  />
                  <div
                    class=" css-j34sbj-control"
                  >
                    <div
                      class=" css-319lph-ValueContainer"
                    >
                      <div
                        class=" css-14el2xx-placeholder"
                        id="react-select-21-placeholder"
                      >
                        Search and select a value...
                      </div>
                      <div
                        class=" css-16cuj0n-Input"
                        data-value=""
                      >
                        <input
                          aria-autocomplete="list"
                          aria-describedby="react-select-21-placeholder"
                          aria-expanded="false"
                          aria-haspopup="true"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          class=""
                          id="selectfield-uniqueitem-0008.search"
                          role="combobox"
                          spellcheck="false"
                          style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                    <div
                      class=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        class=" css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        class=" css-tlfecz-indicatorContainer"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <input
                    name="x.1"
                    type="hidden"
                    value=""
                  />
                </div>
              </div>
            </div>
          </section>
          <div
            class="del-item"
            id="selectfield-uniqueitem-000a.remove"
            role="button"
            tabindex="0"
          >
            <i
              class="fa fa-minus "
            />
            <label />
          </div>
        </li>
        <div
          class="add-item"
          id="selectfield-uniqueitem-000c"
          role="button"
          tabindex="0"
        >
          <i
            class="fa fa-plus "
          />
          <label />
        </div>
      </ul>
    </section>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`<SelectField> <SelectField> - renders a select with correct value (model) 1`] = `
LoadedCheerio {
  "0": <section>
    <div
      class="euiFormRow euiFormRow--fullWidth"
      id="snapshot-test-row"
    >
      <div
        class="euiFormRow__labelWrapper"
      >
        <label
          class="euiFormLabel euiFormRow__label"
          for="snapshot-test"
          id="snapshot-test-label"
        />
         
        <div
          class="euiText euiText--medium"
        />
      </div>
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class=" css-b62m3t-container"
          id="snapshot-test"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-13-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-j34sbj-control"
          >
            <div
              class=" css-319lph-ValueContainer"
            >
              <div
                class=" css-30t56x-singleValue"
              >
                b
              </div>
              <div
                class=" css-16cuj0n-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="snapshot-test.search"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <div
                aria-hidden="true"
                class=" css-tlfecz-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                  />
                </svg>
              </div>
              <span
                class=" css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-tlfecz-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="x"
            type="hidden"
            value="b"
          />
        </div>
      </div>
    </div>
  </section>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
